staging/src/k8s.io/apimachinery/pkg/util/cache/lruexpirecache.go

```diff
@@ -61,29 +53,127 @@ func NewLRUExpireCacheWithClock(maxSize int, clock Clock) *LRUExpireCache {
 func (c *LRUExpireCache) Get(key interface{}) (interface{}, bool) {
+	ce, _ := c.getOrBegin(key, 0)
+	if ce == nil {
+		return nil, false
+	}
+	if ce.wait != nil {
+		return nil, false
+	}
+	return ce.value, true
+}
+
+// ComputeFunc is used by GetOrWait to perform some work. You *must* exit if
+// the abort chan is closed/has a value. On failure, return a ttl of 0 to avoid
+// being cached; this will cause all waiting on the function to observe a cache miss.
+type ComputeFunc func(abort <-chan time.Time) (value interface{}, ttl time.Duration)
+
+// GetOrWait returns the value at the specified key from the cache if it exists
+// and is not expired. Otherwise, it calls the compute() function to produce
+// the value; requests for the same key that arrive within computationTime will
+// be blocked while the computation is in progress; for all such requests,
+// compute() will only be called once.
+//
+// In the unlikely event the cache has so much churn that a freshly computed
+// value can't be regotten, this will attempt to compute it again, but not more
+// than 3 times, and not if more than `computationTime` passes overall.
+func (c *LRUExpireCache) GetOrWait(key interface{}, compute ComputeFunc, computationTime time.Duration) (interface{}, bool) {
+	thisTTL := computationTime
+	if thisTTL < 0 {
+		thisTTL = 0
+	}
+	ce, ch := c.getOrBegin(key, thisTTL)
+	if ch != nil {
+		// We are the first caller for this key, so begin the
+		// computation.
+		abort := c.clock.After(thisTTL)
+		go func() {
+			defer close(ch)
+			value, ttl := compute(abort)
+			if ttl > 0 {
+				// write a new entry before closing the channel
+				// so that new getters of this key can check
+				// ce.value and not wait; existing waiters
+				// won't look at finalValue until after we
+				// close the channel, so there is no race.
+				c.Add(key, value, ttl)
+				ce.finalValue = value
+			} else {
+				c.Remove(key)
+				ce.finalValueOmitted = true
+			}
+		}()
+	}
+	if ce == nil {
+		// This can happen if the computationTime was <= 0, preventing
+		// us from caching even the attempt to do work.
+		return nil, false
+	}
+	if ce.wait == nil {
+		// ce.wait is nil, so there is no need to wait for a
+		// computation to finish.
+		return ce.value, true
+	}
+
+	// We must wait for a computation to finish.
+	<-ce.wait
+	if ce.finalValueOmitted {
+		return nil, false
+	}
+	return ce.finalValue, true
+}
+
+// If ttl <= 0, it will not begin anything and instead return (nil, nil).
+func (c *LRUExpireCache) getOrBegin(key interface{}, ttl time.Duration) (*cacheEntry, chan<- struct{}) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
 	e, ok := c.cache.Get(key)
-	if !ok {
-		return nil, false
+	if ok {
+		n := c.clock.Now()
+		et := e.(*cacheEntry).expireTime
+		if n.After(et) {
+			c.cache.Remove(key)
+			ok = false
+		}
 	}
-	if c.clock.Now().After(e.(*cacheEntry).expireTime) {
-		c.cache.Remove(key)
-		return nil, false
+	if ok {
+		return e.(*cacheEntry), nil
+	}
```

Comment:
nit on the variation: 

```	
	if e, ok := c.cache.Get(key); ok {
		n := c.clock.Now()
		et := e.(*cacheEntry).expireTime
		if n.After(et) {
			c.cache.Remove(key)
		} else {
			return e.(*cacheEntry), nil
		}
	}
```

It reads more like ' 'if we get from the cache and the entry isn't expired return it, but if it is, remove the entry so that no one else gets it and let's pretend nothing happened''

Reply:
cleaned up.